plugins/single_plugins/place.cpp

```diff
@@ -0,0 +1,106 @@
+#include <view.hpp>
+#include <core.hpp>
+#include <workspace-manager.hpp>
+#include <signal-definitions.hpp>
+
+using namespace std;
+
+class wayfire_place_window : public wayfire_plugin_t
+{
+	signal_callback_t created_cb;
+	signal_callback_t workarea_changed_cb;
+	int cascade_x, cascade_y;
+
+	public:
+	void init(wayfire_config *config)
+	{
+		created_cb = [=] (signal_data *data)
+		{
+			auto view = get_signaled_view(data);
+
+			if (view->role != WF_VIEW_ROLE_TOPLEVEL ||
+				view->parent ||view->fullscreen ||
+				view->maximized)
+				return;
+
+			auto workarea = output->workspace->get_workarea();
+			auto section = config->get_section("place");
+			auto placement_mode = section->get_option("mode", "center");
+			string mode = placement_mode->as_string();
+
+			if (mode == string("cascade"))
+				cascade(view, workarea);
+			else if (mode == string("random"))
+				random(view, workarea);
+			else
+				center(view, workarea);
+		};
+
+		workarea_changed_cb = [=] (signal_data *data)
+		{
+			auto workarea = output->workspace->get_workarea();
+			cascade_x = workarea.x;
+			cascade_y = workarea.y;
+		};
+
+		output->connect_signal("reserved-workarea", &workarea_changed_cb);
+		output->connect_signal("map-view", &created_cb);
+	}
+
+	void cascade(wayfire_view &view, wf_geometry workarea)
+	{
+		wf_geometry window = view->get_wm_geometry();
+
+		if (cascade_x + window.width > workarea.x + workarea.width ||
+			cascade_y + window.height > workarea.y + workarea.height)
+		{
+			cascade_x = workarea.x;
+			cascade_y = workarea.y;
+		}
+
+		view->move(cascade_x, cascade_y);
+
+		cascade_x += 50;
```

Comment:
are we sure this should be done this way (hardcoded values)? Maybe add 2-3% of the resolution?

Reply:
It could be done any way, I think the main idea for this primitive placement mode is so windows are not placed directly on top of each other, so the user can see them as they open.

Reply:
but, on a 4k monitor and a low res one the plugin would have a different(IMO) behavior. How does compiz do it or you don't know?